milestones/models.py

```diff
@@ -0,0 +1,123 @@
+which leverages Django's signal framework.
+"""
+
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+
+class Milestone(TimeStampedModel):
+    """
+    A Milestone is a representation of an accomplishment which can be
+    attained by a user. Milestones have a base set of meta data
+    describing the milestone, including id, name, and description.
+    Milestones can be used to drive functionality and behavior 'behind
+    the scenes' in Open edX, such as with the Pre-Requisite Course and
+    Course Entrance Exam use cases.
+    """
+    namespace = models.CharField(max_length=255, db_index=True)
+    name = models.CharField(max_length=255, db_index=True)
+    description = models.TextField()
+    active = models.BooleanField(default=True)
+
+    class Meta:
+        """ Meta class for this Django model """
+        unique_together = (("namespace", "name"),)
+
+
+class MilestoneRelationshipType(TimeStampedModel):
+    """
+    A MilestoneRelationshipType represents a category of link available
+    between a Milestone and a particular learning object (such as a
+    Course). In addition to learning objects, a MilestoneRelationshipType
+    can also represent the link between a Milestone and other platform
+    entities (such as a User).  For example, a Course Author may
+    indicate that Course 101 "fulfills" Milestone A, creating a new
+    CourseMilestone record in the process. When a User completes
+    Course 101, a new UserMilestone record is created reflecting the
+    newly-attained Milestone A.  The Course Author may also indicate
+    that Course 102 "requires" Milestone A, yielding a second
+    CourseMilestone record.  Because the User has gained Milestone A
+    (via Course 101), they can access Course 102.
+
+    This same process of indicating MilestoneRelationshipTypes can be
+    applied to other learning objects as well, such as course content
+    (XBlocks/modules).
+    """
+    name = models.CharField(max_length=255, db_index=True, unique=True)
+    description = models.TextField(blank=True)
+    active = models.BooleanField(default=True)
+
+
+class CourseMilestone(TimeStampedModel):
+    """
+    A CourseMilestone represents the link between a Course and a
+    Milestone. Because Courses are not true Open edX entities (in the
+    Django/ORM sense) the modeling and integrity will be limited to that
+    of specifying CourseKeyFields in this model, as well as related ones
+    below. In addition, a MilestoneRelationshipType specifies the
+    particular sort of relationship that exists between the Course and
+    the Milestone, such as "requires".
+    """
+    course_id = models.CharField(max_length=255, db_index=True)
+    milestone = models.ForeignKey(Milestone, db_index=True)
+    milestone_relationship_type = models.ForeignKey(MilestoneRelationshipType, db_index=True)
+    active = models.BooleanField(default=True)
```

Comment:
Given that `CourseMilestone`, `CourseContentMilestone`, `UserMilestone`, and others in the future all share a bunch of fields (the timestamps, milestone, milestone_relationship_type, active), maybe that could be factored out into an abstract model? That would help with consistency and also give you a convenient place to put common logic like soft-deletes.

So once it's factored out, this class def would look something like:

```python
class CourseMilestone(MilestoneRelationModel):
    course_id = models.CharField(max_length=255, db_index=True)

    class Meta:
        """ Meta class for this Django model """
        unique_together = (("course_id", "milestone"),)
```